### PR TITLE
Resend attendee emails WIP

### PR DESCRIPTION
**Pull request by migetman9**
Hi. @bds002 requested this feature. I think this is what needs to be done to make it happen. I've attempted to test it, but haven't received an email to the attendee email I entered. I figured doing a pull would be the best way to get advice. Thanks!

### CODE REVIEW

File: uber/site_sections/registration.py [hunk 1]
```diff
@@ -167,6 +167,18 @@ def history(self, session, id):
                                .order_by(Tracking.when).all()
         }
 
+    @ajax
+    def resend_email(self, session, id):
+        message = {''}
+        if id:
+            try:
+                email = session.query(Email).filter(Email.model == 'Attendee', Email.fk_id == id).all()[0]
```
Comment (kitsuta) [inline uber/site_sections/registration.py]: Was there a particular reason you didn't use the built-in SQLAlchemy method `first()` instead of accessing the first element of the all() dict?

File: uber/site_sections/registration.py [hunk 2]
```diff
@@ -167,6 +167,18 @@ def history(self, session, id):
...
+                send_email(c.ADMIN_EMAIL, email.dest, email.subject, email.html)
+                message = 'Email Sent!'
+            except ValueError:
```
Comment (kitsuta) [inline uber/site_sections/registration.py]: Although I generally like the try/except block format, you may want to have two checks here: one for the email existing, and one for more generic send_email failures. It _should_ be possible to check for a row in the database without throwing an exception. So you'd have something a bit more like (pseudo-code):
```
email = session.blah.first()
if not email:
  return 'No emails found with that ID'
try:
  send_email(blah)
except:
  return 'Could not resend email. Try again or contact your system administrator.'
```

File: uber/site_sections/registration.py [hunk 3]
```diff
@@ -167,6 +167,18 @@ def history(self, session, id):
...
+                message = 'No Emails Found With That ID'
```
Comment (kitsuta) [inline uber/site_sections/registration.py]: This is purely a stylistic thing, but most of our messages do not use title capitalization and instead read more like sentences (usually without periods at the end).

File: uber/templates/registration/history.html [hunk 4]
```diff
@@ -67,3 +67,25 @@
...
+        var resend = function(email_id){
+            $.ajax(
+                    {
+                        method: 'POST',
+                        url:'resend_email',
+                        data:{
+                            csrf_token: csrf_token,
+                            id: email_id
+                        },
+                        success: function(response){
```
Comment (kitsuta) [inline uber/templates/registration/history.html]: I am not hip to the AJAX, so I am not sure how this function knows whether the function it calls it successful or not when you return a string either way. Also, does this mean that `message` is not ever actually shown to the administrator?

**Comment by kitsuta** [comment]
Awesome @migetman9, thanks for taking the lead on this! :)

One very, _very_ poorly-documented feature of RAMS is that dev boxes* can, in fact, send emails. However, they will only send emails to:
1) The email configured as the 'system administrator' email (for us, this is Eli's email)
2) Email addresses ending with @mailinator.com.

Mailinator.com is a super handy 'public' mailbox - there's no log in or account creation required to receive email at a mailinator.com address. You simply make something up and then go to their website to check that email's inbox. Of course, anyone can do this, so I don't recommend using a generic mailbox name like test@mailinator.com.

Additionally, the system will/should always log emails that it should have sent to any mail address. So even if you use a mailbox that doesn't exist, you _should_ still see the email in question in the /emails/index page.

*Dev boxes are simply any system where the config variable DEV_BOX is set to True. For obvious reasons, this is the default config.

File: uber/site_sections/registration.py (hunk 1 above)
Comment (migetman9) [inline uber/site_sections/registration.py]: forgot I can use first! In my new commit.

File: uber/templates/registration/history.html (hunk 4 above)
Comment (migetman9) [inline uber/templates/registration/history.html]: I am also not hip to the ajax. If the function hits any sort of uncaught error it triggers the error response, if anything is returned it triggers the success response.

Comment (migetman9) [inline uber/templates/registration/history.html]: I am unsure about displaying the message. These are UI things I am planning to solve with @cybersenshi soon™

File: uber/site_sections/registration.py [hunk 5]
```diff
@@ -167,6 +167,23 @@ def history(self, session, id):
...
+    def resend_email(self, session, id, a_id):
+        message = {''}
+        if id:
+            email = session.query(Email).filter(Email.model == 'Attendee', Email.fk_id == id).first()
+            attendee = session.attendee(a_id, allow_invalid=True)
+            if email is not None:
+                try:
+                    send_email(c.ADMIN_EMAIL, attendee.email, email.subject, email.html)
+                    message = 'The email has been resent'
+                except:
```
Comment (migetman9) [inline uber/site_sections/registration.py]: I don't know how to access AmazonError in this except statement

File: uber/site_sections/registration.py [hunk 6]
```diff
@@ -167,6 +167,23 @@ def history(self, session, id):
...
+                    message = "Invalid Key Or Something Maybe"
+                    #Message does nothing currently
+                    #I need to learn Ajax
+
+        return message
```
Comment (migetman9) [inline uber/site_sections/registration.py]: Still currently pointless. Looking into more ajax.

File: uber/templates/registration/history.html [hunk 7]
```diff
@@ -67,3 +67,26 @@
...
+    <script>
+        var resend = function(email_id){
+            $.ajax(
+                    {
+                        method: 'POST',
+                        url:'resend_email',
+                        data:{
+                            csrf_token: csrf_token,
+                            id: email_id,
+                            a_id: '{{ attendee.id }}'
```
Comment (migetman9) [inline uber/templates/registration/history.html]: Without this the email_id contains an outdated email

**Comment by binary1230** [comment]
hey @migetman9 is physically here at the office for the next 2 weeks woo.

I went over an alternative approach to this that instead of calling send_email() directly, we create a new column in the db called 'Email.resend' and if it's true, we resend the email on the next daemon run.

The advantage to this approach is that if email sending goes offline, we've queued up the response to go out, instead of immediately sending it and failing.  And also we could in the future bulk mark a bunch of emails as needing to be re-sent.

I put some notes here for implementation for Dare:
https://github.com/magfest/ubersystem/compare/master...magfest:reset_emails_dom?expand=1

**Comment by bds002** [comment]
Where do we stand with this? 
I am starting to get pinged from DH's about re-sending emails to their staffers to get registered.

**Comment by binary1230** [comment]
Dare and I were working on an alternative approach to this which would better fit in the existing framework (namely instead of calling send_email() directly, we mark the email as needing to be resent and then the existing email system handles resending it).  Not sure what happened to that code though, @migetman9?

**Comment by binary1230** [comment]
ah as dare pointed out, the code that needs to be in this PR is currently in https://github.com/magfest/ubersystem/tree/reset_emails_dom

that code is a good way to do this same thing, but I think needs some extensive testing.

also, this PR will be heavily affected by the merge of https://github.com/magfest/ubersystem/pull/1896 which refactors a bunch of the guts of the email subsystem.  it should be pretty simple to redo though, but will require some careful merging.

**Comment by migetman9** [comment]
If I recall correctly, the work is pretty much done, but Dom never got
around to the final few steps of test and merge.

**Comment by binary1230** [comment]
FOR ANYONE WHO WANTS TO PICK UP THIS FEATURE IN THE FUTURE:

1.  start with this PR.
2.  grab the code from this branch: https://github.com/magfest/ubersystem/tree/reset_emails_dom and merge/combine it in here, it contains slightly more updated concepts.

**Comment by binary1230** [comment]
rob ressurected this from the dead here: https://github.com/magfest/ubersystem/pull/2676
